### PR TITLE
LPD-39115

### PR DESCRIPTION
**Pull request by ChrisKian**
https://liferay.atlassian.net/browse/LPD-39115
https://liferay.atlassian.net/browse/LPD-39986

This PR also contains the test fix for LPD-39986, since the SAML test suite will partially fail without it.

The solution I came up with for the relayState token caching problem was to add a new "relayState" column onto the SamlSpAuthRequest table.  Since this problem only occurs with SP initiated SSO, this approach made the most sense to me.

Please let me know if you have any questions, thanks!

### CODE REVIEW

**Comment by ChrisKian** [comment]
ci:test:relevant

**Comment by ChrisKian** [comment]
ci:test:saml

**Comment by ChrisKian** [comment]
Upgrade test failure look unrelated, since [the error message](https://ff6c79dec8bf92f3bc71305abd225bf9c429c51d1ec5573726d017f-apidata.googleusercontent.com/download/storage/v1/b/testray-results/o/2024-10%2Ftest-1-33%2Ftest-portal-acceptance-pullrequest(master)%2F6890%2Ffunctional-tomcat90-mysql57%2F1%2F13%2FLocalFile.PortalDatabaseValidationUpgrade_CompareLatest7413PartitionedUpgradedAndFreshDBSchemas%2Findex.html.gz?jk=AdvOr8sp4xZ74ibs6AmPpDFz388I9R20JPKpyuIAYtAd2rFveWHJpKkmo8sTHapngtowaBKt3cOPrHb6ULwIBbfdHTztSd0XP-BCHQUTlEsax-sb4fKRp4nKNC5ivQUpdo1twJAvk63PLghx4H7cvxZIkuJG15N1lc2Xi1TEtEqtxkVmW3Odn3uGF4RFtI5mhkqnxm4wbamStoFBz6oz2tPooPSDMBylOavwFjW3PPB2EtfV_Der6NsyNumqyUzn4nUBWQ74AvxQIHEPornj94DX_fJut_naW7zAJzWmMvz4dEMdI1EmYi6Fdlx5NT2sQ2XXieguNed8w3D9OlmXt2hNPQ5be5TK4Cxh4-RrAvLhP8dxLpTxfah25bvJ-x8zexLWBmCzi2soMGADLleU-BN8mPQnLXy_ta1EQEKZI26i_alyQTRCp5WFLRjAiX0jDFPxytOWQWs9jkOpssuxoP84X8WwiPt7lE_ezWDkVPgJ9kWXsinzKdW6lWvOGnQQYJsXNFH0PBw-JnIpLnFAUsLjT5YymP_sXjog6B6d2kcTts_hImINVfv59nJ2faE48-boXVZNETKmgZIIFamcGZV-XMAzXZUAvM_FofDS0AzuSenj9LxuxkD30TwwyELz0F_ZkRp4QEq89lx5VjrQLw64sFEfjB8TNKb2jJ5h9wPv99Wz9gX_cQLcW4ouKX02MuUr-IZEOX0NsR5p83FKSMKg04YR2HYragO9aU7WpX670a2BBtRfi204_ONrH1MUYbcQwspbeheSoRIQEvaxFcYOIXVKKkUjUT-aovfS_6Zvssb72YIzm4enTQ9epZF5CQJDRmeivjnf5xG6KlUPFTrfFDrBy6uJ2XA1PPKkcNPDXKyxWT-WM4AobP9WZHzbQDDgGXW72RqarY11ZEx6qHkpOeE26iZFoVae97QH8Ng2KnDU_ps8sUmLB6OARC3h_6P24tKNav5kSU-JQZP34fjBkTHaZIkkFHVvaqAOboYSoSVetmHzwM3ClIxjRmmkvvHm3Vku3czzxSFRLFrN0dv4CGpL6ayKUpuSb1C_Q5QvpsV9yNdSeFfMyVmBn9mKvHVxQ8OcKY6JF0l-OYPHlpW_mzMqtFeO9CrpnIakrgdPDucKhcCwihbI6bN5m5awbykOoyRJNVVJsAnTf37W6sZE6MvwgICulnPFrrZyA80LxjA36AjtonkiTYqd3jUmiHoA9jxb1i1abUmqjCMjejs-xKVAlh3WHS61VmfEcLSrSgTmfbmqPx2oXz7UhcQhH5mR3lPvn5VhUrmAo47Iz0B7fstSAB_uAjjfmYYyjvkzhWgmclaqCoLstD-kEToIW3XNnSqgBy_8_ex0EDMZEGmyOq2hIn9k-hmVHU1S5qIcR64hDMhxHRToSiwRrQTcKASSZmBNGa4Yszrm&isca=1) doesn't contain the SamlSpAuthRequest table.

Relevant test failure also seem unrelated: 
```
com.liferay.site.exception.InitializationException: com.liferay.portal.kernel.exception.SystemException: com.liferay.portal.kernel.dao.orm.ORMException: javax.persistence.OptimisticLockException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect) : [com.liferay.portlet.documentlibrary.model.impl.DLFileVersionImpl#49407]
```

File: modules/dxp/apps/saml/saml-persistence-service/src/main/java/com/liferay/saml/persistence/service/impl/SamlSpAuthRequestLocalServiceImpl.java [hunk 1]
```diff
@@ -32,7 +32,7 @@ public class SamlSpAuthRequestLocalServiceImpl
 
 	@Override
 	public SamlSpAuthRequest addSamlSpAuthRequest(
-		String samlIdpEntityId, String samlSpAuthRequestKey,
```
[the remark this answers is not in the document]
Comment (ChrisKian) [reply modules/dxp/apps/saml/saml-persistence-service/src/main/java/com/liferay/saml/persistence/service/impl/SamlSpAuthRequestLocalServiceImpl.java:35]: No particular reason, I think I just added it initially and forgot to sort.

samlRelayState is fine

**Comment by stian-sigvartsen** [comment]
Thanks @ChrisKian ! The approach looks good overall, but we should make it less "accidentally working" by completely removing the use of the in-memory cache for the Service Provider role handling if RelayState.

Please see https://github.com/liferay-appsec/liferay-portal/pull/2045

**Pull request closed without merging**